### ISSUE TITLE
TypeError from most spack commands after loading gptune

### ISSUE THREAD

**Issue by wspear** [LLNL/spack#28713]
### Steps to reproduce

After installing gptune like `spack install gptune` in a freshly cloned spack (with a fresh .spack directory) doing 
`spack load gptune`
seems to break spack. Most commands result in `TypeError: __init__() got an unexpected keyword argument 'required'`. Full stack trace from spack debug below.

### Error message

```
wspear@yu:~/bin/SPACK/spack$ spack find gptune
==> Error: __init__() got an unexpected keyword argument 'required'
wspear@yu:~/bin/SPACK/spack$ spack -d find gptune
==> [2022-02-01-08:59:05.044838] Reading config file /home/wspear/bin/SPACK/spack/etc/spack/defaults/config.yaml
Traceback (most recent call last):
  File "/home/wspear/bin/SPACK/spack/bin/spack", line 98, in <module>
    sys.exit(spack.main.main())
  File "/home/wspear/bin/SPACK/spack/lib/spack/spack/main.py", line 900, in main
    return _main(argv)
  File "/home/wspear/bin/SPACK/spack/lib/spack/spack/main.py", line 859, in _main
    command = parser.add_command(cmd_name)
  File "/home/wspear/bin/SPACK/spack/lib/spack/spack/main.py", line 335, in add_command
    self.subparsers = self.add_subparsers(metavar='COMMAND',
  File "/home/wspear/bin/SPACK/spack/lib/spack/spack/main.py", line 314, in add_subparsers
    sp = super(SpackArgumentParser, self).add_subparsers(**kwargs)
  File "/home/wspear/bin/SPACK/spack/opt/spack/linux-ubuntu20.04-westmere/gcc-9.3.0/py-argparse-1.4.0-he26hmfdaw3epehrw4sljthh264y3yp2/lib/python3.8/site-packages/argparse.py", line 1698, in add_subparsers
    action = parsers_class(option_strings=[], **kwargs)
TypeError: __init__() got an unexpected keyword argument 'required'
wspear@yu:~/bin/SPACK/spack$ spack info gptune
==> Error: __init__() got an unexpected keyword argument 'required'
wspear@yu:~/bin/SPACK/spack$ spack find
==> Error: __init__() got an unexpected keyword argument 'required'
wspear@yu:~/bin/SPACK/spack$ spack info
==> Error: __init__() got an unexpected keyword argument 'required'
wspear@yu:~/bin/SPACK/spack$ spack location
==> Error: __init__() got an unexpected keyword argument 'required'
wspear@yu:~/bin/SPACK/spack$ spack help
==> Error: __init__() got an unexpected keyword argument 'required'
wspear@yu:~/bin/SPACK/spack$ spack debug report
==> Error: __init__() got an unexpected keyword argument 'required'
```

### Information on your system

* **Spack:** 0.17.1-1043-0a595d488b
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-westmere
* **Concretizer:** clingo

The script operation for loading gptune:
[gptune-load.txt](https://github.com/spack/spack/files/7980560/gptune-load.txt)

@liuyangzhuan 
@adamjstewart 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by adamjstewart** [comment]
The issue is that `gptune` has an indirect dependency on `py-argparse`. `py-argparse` is a backport of the `argparse` module from Python 2.7 to older versions of Python. Spack also has a copy of `argparse` that is backported from modern Python, or at least hacked to include support for the `required` submodule parameter that was added in Python 3. So because `py-argparse` gets added to the `PYTHONPATH` when you load `gptune`, Spack picks up that module before our vendored copy.

There are two solutions to this problem, both of which we should probably do:

1. Change all `depends_on('py-argparse')` to `depends_on('py-argparse', when='^python@:2.6,3.0:3.1')`, I'll submit a PR to do this.
2. Add `lib/spack/external` to the beginning of `sys.path`, not the end. I think we changed this at one point to allow for users to use newer versions of externals, but this can be problematic. @tgamblin @alalazo would likely know where to fix this
